.github/workflows/merge-to-stage.js

```diff
@@ -0,0 +1,233 @@
+};
+
+const merge = async ({ prs }) => {
+  console.log(`Merging ${prs.length || 0} PRs that are ready... `);
+  for await (const { number, files, html_url, title } of prs) {
+    if (files.some((file) => seen[file])) {
+      await slackNotification(slack.fileOverlap({ html_url, number, title }));
+      continue;
+    }
+    files.forEach((file) => (seen[file] = true));
+    if (!process.env.LOCAL_RUN)
+      await github.rest.pulls.merge({ owner, repo, pull_number: number });
+    body = `- [${title}](${html_url})\n${body}`;
+    await slackNotification(slack.merge({ html_url, number, title }));
+  }
+};
+
+const getStageToMainPR = () =>
+  github.rest.pulls
+    .list({ owner, repo, state: 'open', base: prod })
+    .then(({ data } = {}) => data.find(({ title } = {}) => title === prTitle))
+    .then((pr) => pr && addLabels({ pr, github, owner, repo }))
+    .then((pr) => pr && addFiles({ pr, github, owner, repo }))
+    .then((pr) => {
+      pr?.files.forEach((file) => (seen[file] = true));
+      return pr;
+    });
+
+const openStageToMainPR = async () => {
+  const { data: comparisonData } = await github.rest.repos.compareCommits({
+    owner,
+    repo,
+    base: prod,
+    head: stage,
+  });
+
+  for (const commit of comparisonData.commits) {
+    const { data: pullRequestData } =
+      await github.rest.repos.listPullRequestsAssociatedWithCommit({
+        owner,
+        repo,
+        commit_sha: commit.sha,
+      });
+
+    for (const pr of pullRequestData) {
+      if (!body.includes(pr.html_url))
+        body = `- [${pr.title}](${pr.html_url})\n${body}`;
+    }
+  }
+
+  try {
+    const {
+      data: { html_url, number },
+    } = await github.rest.pulls.create({
+      owner,
+      repo,
+      title: prTitle,
+      head: stage,
+      base: prod,
+      body,
+    });
+    await slackNotification(slack.openedSyncPr({ html_url, number }));
+  } catch (error) {
+    if (error.message.includes('No commits between main and stage'))
+      return console.log('No new commits, no stage->main PR opened');
+    throw error;
+  }
+};
+
+const main = async (params) => {
+  github = params.github;
+  owner = params.context.repo.owner;
+  repo = params.context.repo.repo;
+  currPrNumber = params.context.issue?.number;
+  core = params.core;
+
+  const now = new Date();
+  for (const { start, end } of RCPDates) {
+    if (start <= now && now <= end) {
+      console.log('Current date is within a RCP. Stopping execution.');
+      return;
+    }
+  }
+  try {
+    const stageToMainPR = await getStageToMainPR();
+    console.log('has Stage to Main PR:', !!stageToMainPR);
+    if (stageToMainPR?.labels.some((label) => label.includes(labels.SOTPrefix)))
+      return console.log('PR exists & testing started. Stopping execution.');
+    const prs = await getPRs();
+    await merge({ prs: prs.filter(({ labels }) => isHighPrio(labels)) });
+    await merge({ prs: prs.filter(({ labels }) => !isHighPrio(labels)) });
```

Comment:
We didn't really follow that rule to begin with 😕 

We mentioned we want that to clearly pin-point an offending PR + easy reverts and make it happen by checking that there is no file overlap.

So yes, we allow an arbitary number of PRs per batch, as long as they only touch [unique files](https://github.com/adobecom/milo/pull/2224/files#diff-ed9219c32bdd6c09304ebc81fc40de2930347168af4cb5e03f590c4375925bedR131)

Reply:
I would probably still set some top number there.. like 10. or 7.
I know its unlikely, but very theoretically it could happen that SOTs won't approve for some time and people will keep and keep adding the PRs.

also this makes me think about a next issue.
right now we first 'form' the batch and only then invite SOTs to verify?
is it going to be the same with automation?

Reply:
If the first SOT label is applied, we don't merge anything into stage anymore. If (no) SOTs are not approving for a long time, we'd have to look into that bottleneck and re-plan accordingly 🤔 

> right now we first 'form' the batch and only then invite SOTs to verify?
is it going to be the same with automation?

Yes, unless you got better ideas, such as pre-stage?! which would bring us back to - just test on the branches 😁 it will be the same.

In theory, if we also have the stage->main MERGE workflow should get a little faster, it would be like this:

3/5 SOT approvals...
last SOT comes in and approves, 4/5 approvals there, PR is merged.
15 Seconds later you got your new batch and stage->main PR & the same SOT can instantly start testing and the new batch already has 1/5 approvals.

Reply:
>If the first SOT label is applied, we don't merge anything into stage anymore. 

just realized that this will work good.
no matter how many PRs were in batch, since 1st SOT checked the batch and approved the rest will be testing the same code. all good